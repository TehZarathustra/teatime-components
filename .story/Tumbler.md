## Preview


### size xs

```
<div>
  some text for the baseline
  <Tumbler
    name='tumbler-xs'
    size='xs'/>
  {' '}
  <Tumbler
    disabled={true}
    name='tumbler-xs-d'
    size='xs'/>
</div>
```


### size s

```
<div>
  some text for the baseline
  <Tumbler
    defaultChecked={true}
    name='tumbler-s'
    size='s'/>
  {' '}
  <Tumbler
    defaultChecked={true}
    disabled={true}
    name='tumbler-s-d'
    size='s'/>
</div>
```


### size m

```
<div>
  some text for the baseline
  <Tumbler
    name='tumbler-m'
    size='m'/>
  {' '}
  <Tumbler
    disabled={true}
    name='tumbler-m-d'
    size='m'/>
</div>
```
