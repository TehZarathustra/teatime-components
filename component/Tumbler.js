'use strict';

const { PropTypes } = require('react');
const Tumbler = require('../view/Tumbler');
const simplify = require('../mixin/simplify');

module.exports = simplify(
  Tumbler,
  (styles, { size }) => ({styles: styles[size]}),
  {
    xs: require('../style/tumbler/tumbler-xs.css'),
    s: require('../style/tumbler/tumbler-s.css'),
    m: require('../style/tumbler/tumbler-m.css'),
  },
  {
    size: 'xs',
  },
  {
    size: PropTypes.oneOf([
      'xs',
      's',
      'm',
    ]),
  }
);
