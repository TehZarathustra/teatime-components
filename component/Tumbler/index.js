/* @flow */
import {genericId} from '../../lib/identity';
import {noop} from '../../lib/util';
import React, {Component} from 'react';
import classNames from 'classnames';

import tumblerLarge from './tumbler-large.css';
import tumblerMedium from './tumbler-medium.css';
import tumblerSmall from './tumbler-small.css';

type Props = {
  checked?: boolean,
  className?: string,
  defaultChecked?: boolean,
  disabled: boolean,
  id?: string,
  name: string,
  offText: string,
  onBlur: (event: SyntheticEvent) => void,
  onChange: (event: SyntheticInputEvent, eventData: {checked: boolean, name: string, value: string}) => void,
  onFocus: (event: SyntheticFocusEvent) => void,
  onText: string,
  size: 'l' | 'm' | 's',
  style?: Object,
  styles?: Object,
  value: string,
};

type InputProps = {
  checked?: boolean,
  className: string,
  defaultChecked?: boolean,
  disabled: boolean,
  id: string,
  name: string,
  onBlur: Function,
  onChange: Function,
  onFocus: Function,
  ref: Function,
  type: string,
  value: string,
};

const cssModules = {
  l: tumblerLarge,
  m: tumblerMedium,
  s: tumblerSmall,
};

class Tumbler extends Component {
  props: Props

  static defaultProps: {
    disabled: boolean,
    offText: string,
    onBlur: Function,
    onChange: Function,
    onFocus: Function,
    onText: string,
    size: 'l' | 'm' | 's',
    value: string,
  }

  css: {
    tumbler: string,
    baseline: string,
    shape: string,
    native: string,
    control: string,
    label: string,
    delimiter: string,
  }

  id: string

  _inputElement: HTMLInputElement

  constructor(props: Props) {
    super(props);

    this.css = props.styles || cssModules[props.size];
    this.id = props.id || genericId();

    (this:any)._onChange = this._onChange.bind(this);
    (this:any)._saveRef = this._saveRef.bind(this);
  }

  componentWillReceiveProps(nextProps: Props) {
    this.css = nextProps.styles || cssModules[nextProps.size];
    if (nextProps.id) this.id = nextProps.id || genericId();
  }

  focus() {
    if (this._inputElement) this._inputElement.focus();
  }

  _computeInputProps(): InputProps {
    const {
      checked,
      defaultChecked,
      disabled,
      name,
      onBlur,
      onFocus,
      value,
    } = this.props;
    const {css, id} = this;

    return {
      checked,
      className: css.native,
      defaultChecked,
      disabled,
      id,
      name,
      onBlur,
      onChange: this._onChange,
      onFocus,
      ref: this._saveRef,
      type: 'checkbox',
      value,
    };
  }

  _onChange(e: SyntheticInputEvent) {
    const {name, onChange} = this.props;
    const {checked, value} = e.target;
    onChange(e, {checked, name, value});
  }

  _saveRef(ref: HTMLInputElement) {
    this._inputElement = ref;
  }

  render() {
    const {className, offText, onText, style} = this.props;
    const {css, id} = this;

    const inputProps = this._computeInputProps();

    // why this looks like a pie,
    // see: http://stackoverflow.com/questions/9273016/why-is-this-inline-block-element-pushed-downward

    return (
      <span className={classNames(css.tumbler, className)} style={style}>
        <span className={css.baseline}>
          <span className={css.shape}>
            {this.renderInput(inputProps)}
            <label className={css.control} htmlFor={id}>
              <span className={css.label}>{onText}</span>
              <span className={css.label}>{offText}</span>
              <span className={css.delimiter}>&nbsp;</span>
            </label>
          </span>
        </span>
      </span>
    );
  }

  renderInput(inputProps: InputProps) {
    return (
      <input {...inputProps}/>
    );
  }
}

Tumbler.defaultProps = {
  disabled: false,
  offText: 'Off',
  onBlur: noop,
  onChange: noop,
  onFocus: noop,
  onText: 'On',
  size: 'm',
  value: 'on',
};

export default Tumbler;
