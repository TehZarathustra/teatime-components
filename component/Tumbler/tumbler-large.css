.tumbler
{
  composes: tumbler from './tumbler.css';

  width: 82px;

  font: 15px/32px arial,sans-serif;
}

.baseline
{
  composes: baseline from './tumbler.css';
}

.shape
{
  composes: shape from './tumbler.css';
}

.native
{
  composes: native from './tumbler.css';
}

.control
{
  composes: control from './tumbler.css';

  margin: 0 0 0 -50px;
}

.label
{
  composes: label from './tumbler.css';

  width: 82px;
  padding: 0 0 0 32px;

  line-height: 30px;
}

.label:first-child
{
  margin: 0 -32px 0 0;
  padding: 0 32px 0 0;
}

.delimiter
{
  composes: delimiter from './tumbler.css';

  width: 32px;
  margin: 0 0 0 -82px;

  line-height: 30px;
}
