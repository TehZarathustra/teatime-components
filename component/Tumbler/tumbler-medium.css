.tumbler
{
  composes: tumbler from './tumbler.css';

  width: 76px;

  font: 13px/28px arial,sans-serif;
}

.baseline
{
  composes: baseline from './tumbler.css';
}

.shape
{
  composes: shape from './tumbler.css';
}

.native
{
  composes: native from './tumbler.css';
}

.control
{
  composes: control from './tumbler.css';

  margin: 0 0 0 -48px;
}

.label
{
  composes: label from './tumbler.css';

  width: 76px;
  padding: 0 0 0 28px;

  line-height: 26px;
}

.label:first-child
{
  margin: 0 -28px 0 0;
  padding: 0 28px 0 0;
}

.delimiter
{
  composes: delimiter from './tumbler.css';

  width: 28px;
  margin: 0 0 0 -76px;

  line-height: 26px;
}
