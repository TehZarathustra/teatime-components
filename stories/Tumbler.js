'use strict';

const { action, storiesOf } = require('@kadira/storybook');
const { delimiter } = require('./stories.css');
const React = require('react');
const Tumbler = require('../component/Tumbler');

storiesOf('Tumbler', module)
  .add('preview', () => (
    <div>
      size `xs`{' '}
      <Tumbler
        name='exs'
        onChange={action('change')}
        size='xs'/>{' '}
      <Tumbler
        disabled={true}
        name='dxs'
        onChange={action('change')}
        size='xs'/>

      <hr className={delimiter}/>

      size `s`{' '}
      <Tumbler
        defaultChecked={true}
        name='es'
        onChange={action('change')}
        size='s'/>{' '}
      <Tumbler
        defaultChecked={true}
        disabled={true}
        name='ds'
        onChange={action('change')}
        size='s'/>

      <hr className={delimiter}/>

      size `m`{' '}
      <Tumbler
        name='em'
        onChange={action('change')}
        size='m'/>{' '}
      <Tumbler
        disabled={true}
        name='dm'
        onChange={action('change')}
        size='m'/>
    </div>
  ));
