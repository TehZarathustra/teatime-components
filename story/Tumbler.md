```
<div>
  size `xs` <Tumbler name='exs' size='xs'/> <Tumbler disabled={true} name='dxs' size='xs'/>

  <hr className='delimiter'/>

  size `s` <Tumbler defaultChecked={true} name='es' size='s'/> <Tumbler defaultChecked={true} disabled={true} name='ds' size='s'/>

  <hr className='delimiter'/>

  size `m` <Tumbler name='em' off='No' on='Yes' size='m'/> <Tumbler disabled={true} name='dm' size='m'/>
</div>
```
