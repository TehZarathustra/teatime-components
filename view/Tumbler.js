'use strict';

const React = require('react');
const { Component, PropTypes } = React;
const { bind, noop } = require('../tools/func');
const { generateId } = require('../tools/identity');
const cssModules = require('react-css-modules');

class Tumbler extends Component {
  constructor(props) {
    super(props);

    this.state = {
      id: this.props.id || generateId(),
    };

    bind(this, 'onChange');
  }

  onChange(e) {
    const { checked, value } = e.target;
    this.props.onChange(e, {checked, value});
  }

  render() {
    const { className, off, on, ...o } = this.props;
    const { id } = this.state;

    /**
     * Still there is an issue about controlled and uncontrolled components,
     * related to the input[type="checkbox"] and input[type="radio"].
     * It results in the way controlled components are determined.
     *
     * @see https://github.com/facebook/react/blob/v15.1.0/src/renderers/dom/client/wrappers/ReactDOMInput.js#L171
     * @see https://github.com/facebook/react/issues/6779
     */

    return (
      <div className={className} styleName='wrapper'>
        <input {...o} id={id} styleName='native' onChange={this.onChange} type='checkbox'/>
        <label htmlFor={id} styleName='control'>
          <span styleName='label'>{on}</span>
          <span styleName='label'>{off}</span>
          <span styleName='delimiter'>&nbsp;</span>
        </label>
      </div>
    );
  }
}

Tumbler.defaultProps = {
  off: 'Off',
  on: 'On',
  onChange: noop,
  styles: {},
};

Tumbler.propTypes = {
  checked: PropTypes.bool,
  defaultChecked: PropTypes.bool,
  disabled: PropTypes.bool,
  name: PropTypes.string.isRequired,
  off: PropTypes.string,
  on: PropTypes.string,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onClick: PropTypes.func,
  onContextMenu: PropTypes.func,
  onDoubleClick: PropTypes.func,
  onFocus: PropTypes.func,
  onKeyDown: PropTypes.func,
  onKeyPress: PropTypes.func,
  onKeyUp: PropTypes.func,
  onMouseDown: PropTypes.func,
  onMouseEnter: PropTypes.func,
  onMouseLeave: PropTypes.func,
  onMouseMove: PropTypes.func,
  onMouseOut: PropTypes.func,
  onMouseOver: PropTypes.func,
  onMouseUp: PropTypes.func,
  onTouchCancel: PropTypes.func,
  onTouchEnd: PropTypes.func,
  onTouchMove: PropTypes.func,
  onTouchStart: PropTypes.func,
  styles: PropTypes.object,
  value: PropTypes.string.isRequired,
};

module.exports = cssModules(Tumbler);
