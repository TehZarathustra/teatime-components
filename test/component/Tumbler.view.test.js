'use strict';

import Tumbler from '../../component/Tumbler';
import React from 'react';
import renderer from 'react-test-renderer';

test('defaultProp #size should have a value `m`', () => {
  expect(Tumbler.defaultProps.size).toBe('m');
});

test('Renders correctly with default props', () => {
  const tree = renderer.create(
    <Tumbler
      name='control'>
      Notifications
    </Tumbler>
  );

  expect(tree).toMatchSnapshot();
});

test('Renders correctly with defined props', () => {
  const tree = renderer.create(
    <Tumbler
      checked
      className='outer'
      disabled
      hasLabel={false}
      id='controls-id'
      name='control'
      offText='~~OFF~~'
      onText='~~ON~~'>
      Notifications
    </Tumbler>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
