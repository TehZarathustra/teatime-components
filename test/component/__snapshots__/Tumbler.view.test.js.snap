// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly with default props 1`] = `
<span
  className="tumbler"
  style={undefined}
>
  <span
    className="baseline"
  >
    <span
      className="shape"
    >
      <input
        checked={undefined}
        className="native"
        defaultChecked={undefined}
        disabled={false}
        id="_teatime1"
        name="control"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="checkbox"
        value="on"
      />
      <label
        className="control"
        htmlFor="_teatime1"
      >
        <span
          className="label"
        >
          On
        </span>
        <span
          className="label"
        >
          Off
        </span>
        <span
          className="delimiter"
        >
           
        </span>
      </label>
    </span>
  </span>
</span>
`;

exports[`Renders correctly with defined props 1`] = `
<span
  className="tumbler outer"
  style={undefined}
>
  <span
    className="baseline"
  >
    <span
      className="shape"
    >
      <input
        checked={true}
        className="native"
        defaultChecked={undefined}
        disabled={true}
        id="controls-id"
        name="control"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="checkbox"
        value="on"
      />
      <label
        className="control"
        htmlFor="controls-id"
      >
        <span
          className="label"
        >
          ~~ON~~
        </span>
        <span
          className="label"
        >
          ~~OFF~~
        </span>
        <span
          className="delimiter"
        >
           
        </span>
      </label>
    </span>
  </span>
</span>
`;
